distributed/comm/ucx.py

```diff
@@ -0,0 +1,366 @@
+"""
+:ref:`UCX`_ based communications for distributed.
+
+See :ref:`communcations` for more.
+
+.. _UCX: https://github.com/openucx/ucx
+"""
+import asyncio
+import itertools
+import logging
+import struct
+
+from .addressing import parse_host_port, unparse_host_port
+from .core import Comm, Connector, Listener, CommClosedError
+from .registry import Backend, backends
+from .utils import ensure_concrete_host, to_frames, from_frames
+from ..utils import ensure_ip, get_ip, get_ipv6, nbytes
+
+import ucp
+
+import os
+
+os.environ.setdefault("UCX_RNDV_SCHEME", "put_zcopy")
+os.environ.setdefault("UCX_MEMTYPE_CACHE", "n")
+os.environ.setdefault("UCX_TLS", "rc,cuda_copy")
+
+logger = logging.getLogger(__name__)
+MAX_MSG_LOG = 23
+PORT = 13337
+
+# set in ~/.dask/config.yaml
+# or DASK_DISTRIBUTED__COMM__UCXADDRESS
+_PORT_COUNTER = itertools.count(PORT)
+
+_INITIALIZED = False
+
+
+def _ucp_init():
+    global _INITIALIZED
+
+    if not _INITIALIZED:
+        ucp.init()
+        _INITIALIZED = True
+
+
+# ----------------------------------------------------------------------------
+# Addressing
+# TODO: Parts of these should probably be moved to `comm/addressing.py`
+# ----------------------------------------------------------------------------
+
+
+def _parse_address(addr: str, strict=False) -> tuple:
+    """
+    >>> _parse_address("ucx://10.33.225.160")
+    """
+    if not addr.startswith("ucx://"):
+        raise ValueError("Invalid url scheme {}".format(addr))
+
+    proto, address = addr.split("://", 1)
+    return proto, address
+
+
+def _parse_host_port(address: str, default_port=None) -> tuple:
+    """
+    Parse an endpoint address given in the form "host:port".
+
+    >>> _parse_host_port("10.33.225.160:13337")
+    ("10.33.225.160", 13337)
+    """
+    if address.startswith("ucx://"):
+        _, address = _parse_address(address)
+
+    # if default port is None we select the next port availabe
+    # ucx-py does not currently support random port assignment
+    import random
+
+    default_port = default_port or random.randint(1024, 65000)
+    return parse_host_port(address, default_port=default_port)
+
+
+def _unparse_host_port(host, port=None):
+    return unparse_host_port(host, port)
+
+
+def get_endpoint_address(endpoint):
+    # TODO: ucx-py: 18
```

Comment:
This is https://github.com/Akshay-Venkatesh/ucx-py/issues/61 I think, which is still open.